opt: Add preliminary support for building a memo

rytaft opened the pull request:
This is the first commit of several needed to build a "memo"
structure from a parsed SQL query.  The memo is the primary data
structure in the query optimizer, used for maintaining the forest of
query plan trees that the optimizer considers as part of its search
for the optimal plan.

This commit includes support for building a memo for simple queries
with select, project, and/or inner join operators. It also
includes support for most scalar expressions, which are the building
blocks of scalar projections and filters in the `WHERE` clause.

This commit also includes a struct for maintaining scope, which
is essentially a list of the variable bindings that are visible to
a particular expression in the query tree. This implementation of
scope also enables name resolution and type checking for the parsed
SQL query. As part of the name resolution process, unresolved column
names in the AST are replaced with a `columnProps` object, which is a
unique representation of a column in the query tree.

Future commits will add support for other relational operators
including aggregations and complex joins.

This code is based on code originally written by @petermattis and
@andy-kimball in the opttoy project.

Release note: None

rytaft committed:
opt: Add preliminary support for building a memo

This is the first commit of several needed to build a "memo"
structure from a parsed SQL query. The memo is the primary data
structure in the query optimizer, used for maintaining the forest of
query plan trees that the optimizer considers as part of its search
for the optimal plan.

This commit includes support for building a memo for simple queries
with select, project, and/or inner join operators. It also
includes support for most scalar expressions, which are the building
blocks of scalar projections and filters in the `WHERE` clause.

This commit also includes a struct for maintaining scope, which
is essentially a list of the variable bindings that are visible to
a particular expression in the query tree. This implementation of
scope also enables name resolution and type checking for the parsed
SQL query. As part of the name resolution process, unresolved column
names in the AST are replaced with a columnProps object, which is a
unique representation of a column in the query tree.

Future commits will add support for other relational operators
including aggregations and complex joins.

This code is based on code originally written by @petermattis and
@andy-kimball in the opttoy project.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21672)
<!-- Reviewable:end -->

andy-kimball commented:
Review status: 0 of 16 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UOQMNHVAztSkYi9Zx:-L3UOQMNHVAztSkYi9Zy:bkuacqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L95)):*
> ```Go
> // for executing the SQL query.  After the build process is complete, the memo
> // forest will contain exactly one tree: the logical query plan corresponding
> // to the AST of the original SQL statement with a few simple transformations.
> ```

I'd actually expect the majority of transformations to be normalization transformations rather than exploration transformations. Some of the normalization transformations will be fairly complex, and when you combine that with transitive triggering of patterns, the factory will often build a tree that looks very different from the original SQL statement. So I don't think you can conclude that normalization transformations are simple and exploration transformations are complex - each category contains both simple and complex cases.

---

*[pkg/sql/opt/build/builder.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UPxq2Lum7qXyHHcRq:-L3UPvYMN_rc2cECskpb:b-inepop) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L137)):*
> ```Go
> 	defer func() {
> 		if r := recover(); r != nil {
> 			err = fmt.Errorf("%v", r)
> ```

Is this the usual way in which Cockroach detects and reports "expected" errors? I'd have expected us to add an "error" return value for each buildXXX method, since that is the usual Go way to handle expected errors. Also, I assume we're going to want to attach file/line/position information to semantic errors as well. I assume the AST has that information for us to grab.

---

*[pkg/sql/opt/build/builder.go, line 325 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3US-wPgHvmXmi47k2D:-L3US-wPgHvmXmi47k2E:b-21vsb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L325)):*
> ```Go
> 			b.buildScalar(t.TypedRight(), inScope),
> 		)
> 	case *tree.ParenExpr:
> ```

NIT: add space before

---

*[pkg/sql/opt/build/testdata/build-inner-join, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UX9JFqtjfCB9DFbX6:-L3UX9JFqtjfCB9DFbX7:bb2kx25) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/testdata/build-inner-join#L1)):*
> ```
> exec-raw
> ```

I was expecting more error cases, especially around name scoping. Do you plan to port all the existing Cockroach test cases? Here's an example of an important error case that I assume would already be part of Cockroach's existing suite:

`SELECT * FROM a, (SELECT * FROM b WHERE a.x = b.x)`

---

*[pkg/sql/opt/xform/factory.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UWFZbxHyJeu15ZfVi:-L3UWFZbxHyJeu15ZfVj:b-f5wsy4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/xform/factory.go#L49)):*
> ```Go
> }
> 
> func newFactory(mem *memo, maxSteps int) *Factory {
> ```

I think we should just get rid of newFactory, and instead just use NewFactory. I think we'll always want to construct a new memo each time we create a factory.

---

*[pkg/sql/opt/xform/factory.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UWiyjMWlOqI0yAHv-:-L3UWiyjMWlOqI0yAHv0:bkm6n7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/xform/factory.go#L57)):*
> ```Go
> // (i.e., a view of the designated lowest cost tree in the memo forest)
> // with the given top level group ID and required physical properties.
> func (f *Factory) ExprView(group GroupID, required *PhysicalProps) ExprView {
> ```

I'm a bit wary of exporting this function. Why not just keep it private? Ultimately, we don't really want other packages using the factory directly. That's because in the general case we need to run the optimizer before we can extract the lowest cost tree for the given required properties.

---

*[pkg/sql/opt/xform/memo_group.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UU6ImHMaRaibMdD7y:-L3UU6ImHMaRaibMdD7z:b-33svl5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/xform/memo_group.go#L34)):*
> ```Go
> 	normExprID exprID = 0
> 
> 	// InvalidGroupID is the invalid group.
> ```

NIT: I'd move this up to just after the definition of GroupID, in its own const statement.

---

*[pkg/sql/optbase/catalog.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UULBC7i1WRQcqYn-L:-L3UULBC7i1WRQcqYn-M:bw5ikgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/optbase/catalog.go#L46)):*
> ```Go
> 
> 	// IsHidden returns true if the column is hidden (e.g., there is always a
> 	// hidden column called rowid if there is no primary key on the table).
> ```

I thought these hidden columns were only used during parsing for the NATURAL JOIN case. But you're saying there are other hidden columns as well, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I just gave this a skim in order to keep familiar with the code. Only a few minor comments below.

---

Review status: 0 of 16 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UaDiOpKdUe_VwjNww:-L3UaDiOpKdUe_VwjNwx:bn5br8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L82)):*
> ```Go
> 
> // Builder holds the context needed for building a memo structure from a SQL
> // statement.  Builder.Build() is the top-level function to perform this build
> ```

FYI, two spaces after a period is an anachronism. http://www.slate.com/articles/technology/technology/2011/01/space_invaders.html

---

*[pkg/sql/opt/build/builder.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UPxq2Lum7qXyHHcRq:-L3UcBIyCuCaD8KUSYV3:b-cxytf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L137)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Is this the usual way in which Cockroach detects and reports "expected" errors? I'd have expected us to add an "error" return value for each buildXXX method, since that is the usual Go way to handle expected errors. Also, I assume we're going to want to attach file/line/position information to semantic errors as well. I assume the AST has that information for us to grab.
</blockquote></details>

No, this isn't the usual mechanism. I did it this way in `opttoy` for convenience, but I think it is actually a reasonable pattern for something where success is the common case. Otherwise we litter the code with `if err != nil { return err }`. While I'm used to that pattern as well, I think it is worth discussing if `panic-with-top-level-recover` approach is reasonable in the builder code. Note that there is precedent for this approach in the stdlib. For example, see `encoding/json`.

---

*[pkg/sql/opt/build/scope.go, line 150 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UbtCHKa0VGXDR2sdC:-L3UbtCHKa0VGXDR2sdD:b-4788z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/scope.go#L150)):*
> ```Go
> }
> 
> var _ tree.TypedExpr = &columnProps{}
> ```

I think you already have this assertion in `column_props.go`. Ditto for `VariableExpr`.

---

*[pkg/sql/opt/build/scope.go, line 166 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UblCwogK97A1ilnbN:-L3UblCwogK97A1ilnbO:b-fvyui5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/scope.go#L166)):*
> ```Go
> // Variable is part of the tree.VariableExpr interface. This prevents the
> // column from being evaluated during normalization.
> func (*columnProps) Variable() {}
> ```

Feels strange to have these methods here instead of `column_props.go`. Is there a reason for that organization/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-:-L3Ud2GEr0nq_Fpt6kZt:b-4bbc6n)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 16 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/scope.go, line 166 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UblCwogK97A1ilnbN:-L3UeFdm3wGXffIf1cwE:bz9n3tz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/scope.go#L166)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Feels strange to have these methods here instead of `column_props.go`. Is there a reason for that organization/
</blockquote></details>

I think they should be over in `column_props.go`. It was an oversight on my part not to move them when I created that file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for the reviews!  Still need to look into a couple of comments, but I've addressed most of them.

---

Review status: 0 of 17 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/build/builder.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UaDiOpKdUe_VwjNww:-L3UljiV09PtGKb0rVfu:b-vix4hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L82)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

FYI, two spaces after a period is an anachronism. http://www.slate.com/articles/technology/technology/2011/01/space_invaders.html
</blockquote></details>

Lol ok :) I've removed them.

---

*[pkg/sql/opt/build/builder.go, line 95 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UOQMNHVAztSkYi9Zx:-L3UlutDebj1hmEiBN-A:b-xdh8cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L95)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd actually expect the majority of transformations to be normalization transformations rather than exploration transformations. Some of the normalization transformations will be fairly complex, and when you combine that with transitive triggering of patterns, the factory will often build a tree that looks very different from the original SQL statement. So I don't think you can conclude that normalization transformations are simple and exploration transformations are complex - each category contains both simple and complex cases.
</blockquote></details>

I've changed the wording a bit - do you think this is better?

---

*[pkg/sql/opt/build/builder.go, line 325 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3US-wPgHvmXmi47k2D:-L3Um4CIwJtGaquOqTbB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L325)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: add space before
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/scope.go, line 150 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UbtCHKa0VGXDR2sdC:-L3Um5yqawK6NuO_UC4-:bdcv6kv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/scope.go#L150)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think you already have this assertion in `column_props.go`. Ditto for `VariableExpr`.
</blockquote></details>

Good point!

---

*[pkg/sql/opt/build/scope.go, line 166 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UblCwogK97A1ilnbN:-L3Um7kV0Ur0z1yhMFKL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/scope.go#L166)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think they should be over in `column_props.go`. It was an oversight on my part not to move them when I created that file.
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/testdata/build-inner-join, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UX9JFqtjfCB9DFbX6:-L3UmC_m0jJtehokXXwd:b4c4mhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/testdata/build-inner-join#L1)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I was expecting more error cases, especially around name scoping. Do you plan to port all the existing Cockroach test cases? Here's an example of an important error case that I assume would already be part of Cockroach's existing suite:

`SELECT * FROM a, (SELECT * FROM b WHERE a.x = b.x)`
</blockquote></details>

Good point.  I'll look around for additional relevant test cases to add.   For this first pass I just focused on trying to cover the correct cases.

---

*[pkg/sql/opt/xform/factory.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UWFZbxHyJeu15ZfVi:-L3UmvnMtWAIcLChT1pC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/xform/factory.go#L49)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think we should just get rid of newFactory, and instead just use NewFactory. I think we'll always want to construct a new memo each time we create a factory.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/factory.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UWiyjMWlOqI0yAHv-:-L3Umwi9x4ic2VkSL7SF:bfo2hd9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/xform/factory.go#L57)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'm a bit wary of exporting this function. Why not just keep it private? Ultimately, we don't really want other packages using the factory directly. That's because in the general case we need to run the optimizer before we can extract the lowest cost tree for the given required properties.
</blockquote></details>

I'm using it in `builder_test.go`. Would it be better to change this to `ExprViewString` and just return the string?  That's all I really need for testing...

---

*[pkg/sql/opt/xform/memo_group.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UU6ImHMaRaibMdD7y:-L3UnHXFZdnr8GC7jsSW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/xform/memo_group.go#L34)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I'd move this up to just after the definition of GroupID, in its own const statement.
</blockquote></details>

Done.

---

*[pkg/sql/optbase/catalog.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UULBC7i1WRQcqYn-L:-L3UnIR2QWy4rp3xmS5f:b-icuuir) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/optbase/catalog.go#L46)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I thought these hidden columns were only used during parsing for the NATURAL JOIN case. But you're saying there are other hidden columns as well, right?
</blockquote></details>

Yea - I'm aware of the one case I mentioned here -- `rowid` -- but there may be others.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-:-L3Uo2GeAC8bDmIfmF9H:b-lrfukb)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
LGTM, just some minor nits (feel free to ignore any of them).

---

Review status: 0 of 17 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 621 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3WsHRAkvbwGByH-Ox0:-L3WsHRAkvbwGByH-Ox1:biqkq4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/builder.go#L621)):*
> ```Go
> }
> 
> var _ tree.IndexedVarContainer = &Builder{}
> ```

similar to scope, a comment explaining when the Builder is used as a container would be good

---

*[pkg/sql/opt/build/builder_test.go, line 182 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3WpvUIP2rl3zkdfLPn:-L3WpvUIP2rl3zkdfLPo:bjpk3qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/builder_test.go#L182)):*
> ```Go
> 
> // parseTypes parses a list of types.
> func parseTypes(colStrs []string) ([]types.T, error) {
> ```

[nit] if we're going to have similar tests in multiple packages, these should be in some testutils package

---

*[pkg/sql/opt/build/column_props.go, line 53 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3Vl0a84YIPPt3VSyvy:-L3Vl0a84YIPPt3VSyvz:b-9fup2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/column_props.go#L53)):*
> ```Go
> }
> 
> func (c columnProps) matches(tblName optbase.TableName, colName optbase.ColumnName) bool {
> ```

[nit] could use a comment documenting the empty tabalename behavior

---

*[pkg/sql/opt/build/scope.go, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3V4ZJc2KJEHpFkQzxy:-L3V4ZJc2KJEHpFkQzxz:bgp2278) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/scope.go#L60)):*
> ```Go
> }
> 
> // VisitPre is part of the Visitor interface.
> ```

[nit] it would be good to add a `var _ tree.Visitor = &scope{}` and a comment explaining what the visitor does and when/why it is used.

---

*[pkg/sql/opt/build/scope.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3WrgmspgUWP9_QQwwQ:-L3WrgmspgUWP9_QQwwR:blzzto7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/scope.go#L105)):*
> ```Go
> }
> 
> var _ tree.IndexedVarContainer = &scope{}
> ```

[nit] it would be good to explain in a comment why this is useful (when do we use `scope` as a container?)

---

*[pkg/sql/opt/build/scope.go, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3VlElSBSKOOwbrnnaV:-L3VlElVw9jD-L760o4R:b4f3vp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/scope.go#L110)):*
> ```Go
> func (s *scope) IndexedVarEval(idx int, ctx *tree.EvalContext) (tree.Datum, error) {
> 	unimplemented("scope.IndexedVarEval")
> 	return nil, fmt.Errorf("unimplemented: scope.IndexedVarEval")
> ```

seems like just using `panic()` would be less tedious

---

*[pkg/sql/opt/build/testdata/build-project, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3VmITnx0fSUkBgrf__:-L3VmITnx0fSUkBgrf_a:b-wy84zk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/testdata/build-project#L110)):*
> ```
> 
> exec-raw
> DROP DATABASE t
> ```

is this necessary? usually logictests don't clean up db state

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-:-L3WqUx7L2NzIzwitOpd:b-yze4jj)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 17 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UPxq2Lum7qXyHHcRq:-L3YvxWZPhDVVw36L-nG:b169a7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L137)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

No, this isn't the usual mechanism. I did it this way in `opttoy` for convenience, but I think it is actually a reasonable pattern for something where success is the common case. Otherwise we litter the code with `if err != nil { return err }`. While I'm used to that pattern as well, I think it is worth discussing if `panic-with-top-level-recover` approach is reasonable in the builder code. Note that there is precedent for this approach in the stdlib. For example, see `encoding/json`.
</blockquote></details>

But how do we distinguish between semantic errors that need to be reported to the user and "real" panics that indicate a bug that we want to handle differently (like wrapping it with an "internal error, please report bug" message)?

---

*[pkg/sql/opt/build/scope.go, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3VlElSBSKOOwbrnnaV:-L3YwgeYraltSk4odUTe:b8ten36) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/scope.go#L110)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

seems like just using `panic()` would be less tedious
</blockquote></details>

I agree that the unimplemented and fatalf helpers seem to cause as much trouble as they save, since they're not factored into escape analysis.

---

*[pkg/sql/opt/xform/factory.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UWiyjMWlOqI0yAHv-:-L3Yx73d7fulGGkYRoWh:b-cipyho) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/xform/factory.go#L57)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I'm using it in `builder_test.go`. Would it be better to change this to `ExprViewString` and just return the string?  That's all I really need for testing...
</blockquote></details>

Ah, right, I was thinking the test was in the same package. In this case, just get rid of the required argument, because the factory will always build the normalized tree that has a single normalized path through it (which minPhysPropsID will access). If callers want a different set of physical props, they'll have to invoke the optimizer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 17 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UPxq2Lum7qXyHHcRq:-L3Z18sGHs3evTO97p4v:b-raz36x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L137)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

But how do we distinguish between semantic errors that need to be reported to the user and "real" panics that indicate a bug that we want to handle differently (like wrapping it with an "internal error, please report bug" message)?
</blockquote></details>

`panic` and `recover` pass an `interface{}`. We can put whatever we want into that `interface{}`. For example, errors can pass the concrete type `builderError` and everything else gets re-paniced. `encoding/json` takes the opposite approach and looks for `runtime.Error` panics:

```
	defer func() {
		if r := recover(); r != nil {
			if _, ok := r.(runtime.Error); ok {
				panic(r)
			}
			err = r.(error)
		}
	}()
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 17 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UPxq2Lum7qXyHHcRq:-L3Z2FIZ2J_ElvSXoSsR:bsermcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L137)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`panic` and `recover` pass an `interface{}`. We can put whatever we want into that `interface{}`. For example, errors can pass the concrete type `builderError` and everything else gets re-paniced. `encoding/json` takes the opposite approach and looks for `runtime.Error` panics:

```
	defer func() {
		if r := recover(); r != nil {
			if _, ok := r.(runtime.Error); ok {
				panic(r)
			}
			err = r.(error)
		}
	}()
```
</blockquote></details>

OK, seems like a reasonable approach.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 17 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UPxq2Lum7qXyHHcRq:-L3Z3I_OxA75enWE49Xn:b-4tpm2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L137)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

OK, seems like a reasonable approach.
</blockquote></details>

Btw, I haven't had any issues with this error handling approach in `opttoy`, though I'll note that the reason this is feasible in this code is because we're not updating shared state and not manipulating locks. We'll need to pay attention if that changes. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 0 of 15 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/opt/build/builder.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UPxq2Lum7qXyHHcRq:-L3ZjSPJKsPogFbrF_jn:b8eke72) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/build/builder.go#L137)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Btw, I haven't had any issues with this error handling approach in `opttoy`, though I'll note that the reason this is feasible in this code is because we're not updating shared state and not manipulating locks. We'll need to pay attention if that changes. 
</blockquote></details>

I've added the `builderError` type check along with a cautionary comment.  Let me know if you think this is sufficient. Thanks!

---

*[pkg/sql/opt/build/scope.go, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3VlElSBSKOOwbrnnaV:-L3ZjKayYfriZWxoAmSN:b-3jxvai) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/scope.go#L110)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I agree that the unimplemented and fatalf helpers seem to cause as much trouble as they save, since they're not factored into escape analysis.
</blockquote></details>

Done. I've removed those helpers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once you've addressed any of @radu's comments that you decide to address.

---

Review status: 0 of 15 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-:-L3_E-B1Y_C0BpoAujiw:b-188v3c)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 15 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 202 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3_oCmoen4C4owKmAbB:-L3_oCmoen4C4owKmAbC:b-nymage) ([raw file](https://github.com/cockroachdb/cockroach/blob/8617f9aa5074c273596ede6ab67661766bef0d78/pkg/sql/opt/build/builder.go#L202)):*
> ```Go
> 
> 	default:
> 		panic(newBuilderErrorf("unexpected statement: %T", stmt))
> ```

I wonder if this should be shortened to `errorf(...)` which internally does the `panic(newBuilderErrorf(...))` dance.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 15 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 202 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3_oCmoen4C4owKmAbB:-L3_qic-Wdx4VdFeNw1X:b-5wxt2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/8617f9aa5074c273596ede6ab67661766bef0d78/pkg/sql/opt/build/builder.go#L202)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I wonder if this should be shortened to `errorf(...)` which internally does the `panic(newBuilderErrorf(...))` dance.  
</blockquote></details>

The problem with that is that then the Go compiler doesn't do escape analysis, and you have to return dummy nil values all over the place.  I think this is better, though maybe we could shorten newBuilderErrorf somehow?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 15 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 202 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3_oCmoen4C4owKmAbB:-L3_w8CZgLbdM4QGdlAG:b-vk4q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8617f9aa5074c273596ede6ab67661766bef0d78/pkg/sql/opt/build/builder.go#L202)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

The problem with that is that then the Go compiler doesn't do escape analysis, and you have to return dummy nil values all over the place.  I think this is better, though maybe we could shorten newBuilderErrorf somehow?
</blockquote></details>

Yeah, that's an irritation. Ok, how about `panic(errorf(...))`. No need for the verbosity of `newBuilderErrorf`. 

Is the problem with the Go compiler or one of the lint checks? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for the comments!  I think I've addressed all of them except for adding the additional test cases.  I'll do that in the next PR unless you think it's important for this one.

---

Review status: 0 of 18 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/opt/build/builder.go, line 621 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3WsHRAkvbwGByH-Ox0:-L3a1n88bA1c7-7gYBzG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/builder.go#L621)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

similar to scope, a comment explaining when the Builder is used as a container would be good
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/builder.go, line 202 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3_oCmoen4C4owKmAbB:-L3a27GTMWpe3TSfmGMJ:bem12et) ([raw file](https://github.com/cockroachdb/cockroach/blob/8617f9aa5074c273596ede6ab67661766bef0d78/pkg/sql/opt/build/builder.go#L202)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yeah, that's an irritation. Ok, how about `panic(errorf(...))`. No need for the verbosity of `newBuilderErrorf`. 

Is the problem with the Go compiler or one of the lint checks? 
</blockquote></details>

Done. I think it's the compiler.

---

*[pkg/sql/opt/build/builder_test.go, line 182 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3WpvUIP2rl3zkdfLPn:-L3a1otLO3GA4zXMTnrC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/builder_test.go#L182)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] if we're going to have similar tests in multiple packages, these should be in some testutils package
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/column_props.go, line 53 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3Vl0a84YIPPt3VSyvy:-L3a1smSwKBKysPRmEol:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/column_props.go#L53)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] could use a comment documenting the empty tabalename behavior
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/scope.go, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3V4ZJc2KJEHpFkQzxy:-L3a1vopyLSEMKbhaGm8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/scope.go#L60)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] it would be good to add a `var _ tree.Visitor = &scope{}` and a comment explaining what the visitor does and when/why it is used.
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/scope.go, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3WrgmspgUWP9_QQwwQ:-L3a1xmeB9aLXK_zG94s:btuotpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/scope.go#L105)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] it would be good to explain in a comment why this is useful (when do we use `scope` as a container?)
</blockquote></details>

On second thought, I think we don't actually need scope to be a container.

---

*[pkg/sql/opt/build/testdata/build-project, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3VmITnx0fSUkBgrf__:-L3ZolWPREkL0KcuNNv1:bkqwkkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a730d562d0e9d0b91cc4b9f89ce97298b4b2a7c5/pkg/sql/opt/build/testdata/build-project#L110)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

is this necessary? usually logictests don't clean up db state
</blockquote></details>

Removed - good point!

---

*[pkg/sql/opt/xform/factory.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-L3UWiyjMWlOqI0yAHv-:-L3Zojrm7-LyQtXOq9Kv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba50389852c1079f3461acce309013a82d22da4a/pkg/sql/opt/xform/factory.go#L57)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Ah, right, I was thinking the test was in the same package. In this case, just get rid of the required argument, because the factory will always build the normalized tree that has a single normalized path through it (which minPhysPropsID will access). If callers want a different set of physical props, they'll have to invoke the optimizer.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21672#-:-L3a2E2-KDhCBLu8fIDO:b-1yamu4)*
<!-- Sent from Reviewable.io -->
